File: src/stacks/impl/StackUsingLinkedList.java
package src.stacks.impl;

import src.linkedlist.ListNode;
import src.stacks.IStack;

public class StackUsingLinkedList<T> implements IStack<T> {
    ListNode<T> head;

    public StackUsingLinkedList() {
        this.head = null;
    }

    @Override
    public void push(T item) {
        ListNode<T> newNode = new ListNode<>(item);
        if (isEmpty()) {
            head = newNode;
            return;
        }

        newNode.next = head;
        head = newNode;
    }

    @Override
    public T pop() {
        if (isEmpty()) return null;

        T item = head.val;
        head = head.next;
        return item;
    }

    @Override
    public T peek() {
        if (isEmpty()) return null;
        return head.val;
    }

    @Override
    public boolean isEmpty() {
        return head == null;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();

        ListNode<T> curr = head;
        if (curr != null) {
            sb.append(curr.val);
            curr = curr.next;
            while (curr != null) {
                sb.insert(0, curr.val + ", ");
                curr = curr.next;
            }
        }

        sb.insert(0, "[");
        sb.append("]");
        return sb.toString();
    }
}

File: src/stacks/impl/StackUsingLinkedList.kt
package src.stacks.impl

import src.linkedlist.ListNode
import src.stacks.IStack

class StackUsingLinkedList<T> : IStack<T> {

    private var head: ListNode<T>? = null

    override fun push(item: T) {
        val newNode = ListNode(item)
        if (isEmpty()) {
            head = newNode
            return
        }
        newNode.next = head
        head = newNode
    }

    override fun pop(): T? {
        if (isEmpty()) return null
        val item = head?.`val`
        head = head?.next
        return item
    }

    override fun peek(): T? = if (isEmpty()) null else head?.`val`

    override fun isEmpty(): Boolean = head == null

    override fun toString(): String = buildString {
        var curr = head
        if (curr != null) {
            append(curr.`val`)
            curr = curr.next
            while (curr != null) {
                insert(0, curr.`val`.toString() + ", ")
                curr = curr.next
            }
        }
        insert(0, "[")
        append("]")
    }
}
